[KEP-2724] Add cross-podset TAS scheduling documentation (#5895)
kubernetes-sigs/kueue: keps/2724-topology-aware-scheduling/README.md

to omit the annotation `kueue.x-k8s.io/podset-slice-size`, as the default
value for it in case of JobSet is `parallelism`.

##### LeaderWorkerSet

According to [Story 6](#story-6) we noticed that some users would like to
co-locate Leader with Workers within the same replica in LeaderWorkerSet.

To allow user to specify the requested topology for leader and workers we 
are adding an annotation `kueue.x-k8s.io/leaderworkerset-group-required-topology`.

**Example**:

```yaml
apiVersion: leaderworkerset.x-k8s.io/v1
kind: LeaderWorkerSet
metadata:
  name: tas-example-lws
  labels:
    kueue.x-k8s.io/queue-name: user-queue
  annotations:

>     kueue.x-k8s.io/leaderworkerset-group-required-topology: "cloud.provider.com/topology-rack"

Reviewer: Why do we need to have LWS dedicated annotation? Could we have more generic one in the following?

```yaml
apiVersion: leaderworkerset.x-k8s.io/v1
kind: LeaderWorkerSet
metadata:
  annotations:
    kueue.x-k8s.io/podset-required-topology: "cloud.provider.com/topology-rack"
    kueue.x-k8s.io/podset-slice-group-name: "leader-worker" // This is a new annotation and value is arbitrary one.
  spec:
    leaderWorkerTemplate:
      leaderTemplate:
        metadata:
          kueue.x-k8s.io/podset-slice-group-name: "leader-worker" // This is a new annotation and the value must be the same value as the top-level metadata.
      workerTemplate:
        metadata:
          kueue.x-k8s.io/podset-slice-group-name: "leader-worker" // This is a new annotation and the value must be the same value as the top-level metadata.
```

Reviewer 2: I'm not yet sure if we need to support other frameworks - we don't have such a use-case yet. The LWS is quite special actually.

However, even if we need at some point support other frameworks, then I think it makes sense still to have `kueue.x-k8s.io/leaderworkerset-group-required-topology` for the ease of use, because otherwise we need to require from users setting two annotations. So, I think introducing the dedicated annotation is not wasted effort anyway.

Finally, ensuring the functionality works properly for all frameworks with two new annotations means more code, more validation, and more testing, and so putting the new feature under risk for 0.13.

Reviewer 2: Yes, I see value in the generic support. 

Still, there is an extra validation cost to ensure all the individual `kueue.x-k8s.io/podset-group-required-topology`, which adds more risk for 0.13 timeline.

Finally, the approach is repetitive from user's perspective. For example the user of LWS needs to put 4 annotations instead of one:
- on leader `kueue.x-k8s.io/podset-group-name: lws-group`, and `kueue.x-k8s.io/podset-group-required-topology: host`
- and the same two annotations on worker

So, I think it still makes sense to have `kueue.x-k8s.io/leaderworkerset-group-required-topology` for the ease of use. Also, the user does not really need to think how to name the PodSetGroup.

Reviewer: As my initial thought, I would support only LWS in the next v0.13. After that, we can expand the supported framework step by step. So, the delaying release schedule risk could be minimized.

Regarding validation costs, I think that it's not too difficult since we can just iterate LWS and detect violations.
We already have more complex scheduling and assignment mechanism in Kueue 😉 

Btw, Despite the proposed `kueue.x-k8s.io/podset-group-name` and `kueue.x-k8s.io/podset-group-required-topology` in my comment, I actually use `kueue.x-k8s.io/podset-group-name`, `kueue.x-k8s.io/podset-group-required-topology`, and `kueue.x-k8s.io/podset-group-member-count` in the example manifest. However, I'm not confident the `kueue.x-k8s.io/podset-group-member-count` would be helpful since I'm ok with introducing only `kueue.x-k8s.io/podset-group-name` and `kueue.x-k8s.io/podset-group-required-topology`, initially. When we expand this feature to other frameworks, we can reconsider the `kueue.x-k8s.io/podset-group-member-count`.

Reviewer: > sgtm, but +1 for following up with support for other frameworks.

yes, we should focus on LWS support for now.

Reviewer: My proposal is that introducing `kueue.x-k8s.io/podset-group-name` and `kueue.x-k8s.io/podset-group-required-topology` and not introduce `kueue.x-k8s.io/leaderworkerset-group-required-topology`.

However, we support cross podset TAS scheduling only for LWS in 0.13.

Reviewer 2: wdyt @tenzen-y , am I reading your message in https://github.com/kubernetes-sigs/kueue/pull/5895#discussion_r2205143212 that long term you only want to have `kueue.x-k8s.io/podset-group-name` and `kueue.x-k8s.io/podset-group-required-topology`?

Reviewer 2: If we think using the two annotations is better approach, but don't have time to implement for other workloads than LWS then I think it is better to leave it undone. For better feedback we could just return validation errors saying Not supported Yet.

TAS is still alpha so not everything needs to be perfect. However if we introduce the interim LWS annotation it will mean extra effort when we break the API in future, for users who start to use it.

Reviewer: My intention is to support only `kueue.x-k8s.io/podset-group-name` and `kueue.x-k8s.io/podset-group-required-topology` for both long and short terms.

And I think that only cross-podset support for LWS is must to have in v0.13. Other cross-podset support for other frameworks is nice to have in v0.13, which means not REQUIRED features for v0.13.

Reviewer 2: Ok, let's go with just `kueue.x-k8s.io/podset-group-name` and `kueue.x-k8s.io/podset-group-required-topology`.

>     kueue.x-k8s.io/leaderworkerset-group-required-topology: "cloud.provider.com/topology-rack"

Reviewer: Just for confirmation about LWS. Are these top-level annotations and labels propagated to the underlying Pods?

  // For example, in the context of JobSet this value is read from jobset.sigs.k8s.io/replicatedjob-replicas.
  SubGroupCount *int32

> 	// PodSetGroup indicates the name of the group of PodSets to which this PodSet belongs to.

Reviewer 2: nit align indentation

  // PodSets with the same `PodSetGroup` should be assigned the same ResourceFlavor
  //
  // +optional

>   PodSetGroup *string `json:"podSetGroup,omitempty"`

Reviewer: In the case of above LWS example, what is the expected name in this podSetGroup?

Author: It can be anything as long as it is the same value on other PodSets you would like to group. It will be set by `leaderworkerset_reconciler` (see here: https://github.com/kubernetes-sigs/kueue/pull/5878/files) to `leader-worker-set-group` if the annotation is present (if it is not, it will be left as nil)

>   // PodSetGroup indicates the name of the group of PodSets to which this PodSet belongs to.
>   // PodSets with the same `PodSetGroup` should be assigned the same ResourceFlavor
>   //
>   // +optional
>   PodSetGroup *string `json:"podSetGroup,omitempty"`

Reviewer: ```suggestion
  // PodSetGroupName indicates the name of the group of PodSets to which this PodSet belongs to.
  // PodSets with the same `PodSetGroupName` should be assigned the same ResourceFlavor
  //
  // +optional
  PodSetGroupName *string `json:"podSetGroupName,omitempty"`
```

It would be better to use "name" suffix since this field indicates a name of PodSerGroup instead of PodSetGroup related parameters.

It is worth noting that the tight fit mentioned above does not guarantee that no free capacity will be left within the assigned domains.

### Cross-PodSet scheduling

In consideration of a [Story 6](#story-6) a cross-podset scheduling is required,
due to the fact that leader and workers are separate PodSets. To be able to co-locate

> leaders with its workers, two mechanisms have to be introduced:

Reviewer 2: nit, maybe "we divide the problem into subproblems:"

as the bullets below are more like subproblems thatn mechanism directly.

Author: Good suggestion. I have rephrased the sentence.

Reviewer 2: sgtm

- Ensure leader and workers end up on the same flavor

> - Find topology domain to co-locate leader and workers

Reviewer: What exactly mean of cross-podset scheduling? Does this indicate "cross-podset topology assignment"?
Or does this mean the "cross-podset flavor assignment"?

Reviewer: Thank you for rephrasing that.

If we want to allow it cross-podset flavor assignment, we need to introduce a more complex flavor assignment algorithm because some preemption and flavor assignment features depend on the fact that each PodSet flavor assignment does not care about other PodSet flavor assignments.

So, I have the below questions for LWS example (story 6):

1. Which scheduling policy flavorFungibility and Cross-PodSet colocation does Kueue respect? If flavor assigner can not assign the same flavor to leader-worker group, what if flavorFungibility effectiveness? Does the flavor assigner select TryNext mode when a leader-worker PodSet group can not be accomdatetd to the same flavor?
2. What if either replica pod is evicted or preempted? The LWS is built on top of Pod integration. So, either Pod could be preempted by others. For example, what if only leader Pod is preempted by others? Will TAS evict worker Pods and try to reschedule an entire PodGroup to the same Topology / Flavor?

Reviewer 2: Maybe @lchrzaszcz can elaborate more, here is my understanding:

> Which scheduling policy flavorFungibility and Cross-PodSet colocation does Kueue respect?

Kueue respects flavorFungibility, but the unit of that is "PodSetGroup", not a single PodSet. 

> If flavor assigner can not assign the same flavor to leader-worker group, what if flavorFungibility effectiveness? 

Then it marks the mode as NoFit, and does not schedule the workload. Even if the individual PodSets, when considered independently could have "Fit".

> Does the flavor assigner select TryNext mode when a leader-worker PodSet group can not be accomdatetd to the same flavor?

Yes.

> What if either replica pod is evicted or preempted?

By preempted you mean preempted by Kueue scheduler (1) , or  kube-scheduler (2)? If:
1. Kueue always preempts entire workloads currently, so the entire LWS group is preempted. It does not matter if they are in the same or different flavors.
2. It should not happen if Kueue controls all workloads, so is corner case. In that case, the preempted Pod is pending, while the rest remains "Running". Same thing will happen for Jobs without this feature.

> The LWS is built on top of Pod integration. So, either Pod could be preempted by others.

The unit of preemption for Kueue is  Workload, regardless if using Pod integration, or any other.

> For example, what if only leader Pod is preempted by others? 

Again, depending which scheduler initiates the preemption. If Kueue, then both leader and workers get preempted. It does not matter if they are in the same or different flavors.

> Will TAS evict worker Pods and try to reschedule an entire PodGroup to the same Topology / Flavor?

Yes, but this is not even related to TAS, but how Kueue handles evictions - the unit of eviction is always entire workload, regardless is quota-based or TAS-based preemption. 

We are considering partial preemption, but this is a new enhancement.

Reviewer: > > Which scheduling policy flavorFungibility and Cross-PodSet colocation does Kueue respect?
> 
> Kueue respects flavorFungibility, but the unit of that is "PodSetGroup", not a single PodSet.

Before this feature, a flavor assignment primitive was a PodSet. However, in this feature, that is a PodSetGroup.
This is an important change. So, we want to mention that in the proposal.

> > If flavor assigner can not assign the same flavor to leader-worker group, what if flavorFungibility effectiveness?
> 
> Then it marks the mode as NoFit, and does not schedule the workload. Even if the individual PodSets, when considered independently could have "Fit".
> > Does the flavor assigner select TryNext mode when a leader-worker PodSet group can not be accomdatetd to the same flavor?
> 
> Yes.

As I mentioned, could you add this new primitive in the proposal? Additionally, is the primitive change only for flavorFungibility?

>> What if either replica pod is evicted or preempted?
>
> By preempted you mean preempted by Kueue scheduler (1) , or kube-scheduler (2)? If:

>> Kueue always preempts entire workloads currently, so the entire LWS group is preempted. It does not matter if they are in the same or different flavors.
It should not happen if Kueue controls all workloads, so is corner case. In that case, the preempted Pod is pending, while the rest remains "Running". Same thing will happen for Jobs without this feature.
The LWS is built on top of Pod integration. So, either Pod could be preempted by others.
>
> The unit of preemption for Kueue is Workload, regardless if using Pod integration, or any other integration.

>> For example, what if only leader Pod is preempted by others?
> 
> Again, depending which scheduler initiates the preemption. If Kueue, then both leader and workers get preempted. It does not matter if they are in the same or different flavors.

That makes sense.

>> Will TAS evict worker Pods and try to reschedule an entire PodGroup to the same Topology / Flavor?
> 
> Yes, but this is not even related to TAS, but how Kueue handles evictions - the unit of eviction is always entire workload, regardless is quota-based or TAS-based preemption.
> 
> We are considering partial preemption, but this is a new enhancement.

That sounds good to me. Could you add this scenario to NonGoal?

@lchrzaszcz If Michal's answer is correct, could you address my comments?

Reviewer: Regarding Preemption and Eviction, as I checked LWS and StatefulSet integration impls again, those are built on top of Pod Group integrations. So, the preemption and eviction will be executed againt those resources as a unit.

However, Deployment is built on top of Pod integration, which means an individual Pod could be evicted / preempted.

In other words, LWS and StatefulSet are compatible with this feature. But, Deployment are not compatible with this feature. However, as we discussed in other thread, we will focus on LWS support for now. So, I'm ok with postponing the investigation if there are any problems with Deployment integrations.

Reviewer: Thank you for the update. I will recheck this proposal.

#### Ensure leader and workers end up on the same flavor

To ensure leader and workers are assigned the same flavor a notion of PodSetGroup is introduced:

```golang
type PodSetTopologyRequest struct {
  // ...

	// PodSetGroup indicates the name of the group of PodSets to which this PodSet belongs to.
	// PodSets with the same `PodSetGroup` should be assigned the same ResourceFlavor
	//
	// +optional
	PodSetGroup *string `json:"podSetGroup,omitempty"`
  
  // ...
}
```

This field specifies the name of a group of PodSets that should be placed on the same flavor.

> This field is optionl and if a PodSet does not define it, it will be placed on a flavor 

Reviewer 2: nit optional

Author: Fixed typo. Thanks for spotting!

It is worth noting that the tight fit mentioned above does not guarantee that no free capacity will be left within the assigned domains.

### Cross-PodSet Topology Aware scheduling

In consideration of a [Story 6](#story-6) a cross-podset topology aware scheduling
is required, due to the fact that leader and workers are separate PodSets. To be able 
to co-locate leaders with its workers in the same topology domain, we divide the
problem into subproblems:

- Ensure leader and workers end up on the same flavor
- Find topology domain to co-locate leader and workers

#### Ensure leader and workers end up on the same flavor

To ensure leader and workers are assigned the same flavor a notion of `PodSet Group` is 

> introduced, indicated by the PodSetTopologyRequest's PodSetGroupName field (see [Internal APIs](Internal APIs)):

Reviewer 2: ```suggestion
introduced, indicated by the PodSetTopologyRequest's PodSetGroupName field (see [Internal APIs](#internal-apis)):
```

It is worth noting that the tight fit mentioned above does not guarantee that no free capacity will be left within the assigned domains.

### Cross-PodSet Topology Aware scheduling

In consideration of a [Story 6](#story-6) a cross-podset topology aware scheduling
is required, due to the fact that leader and workers are separate PodSets. To be able 
to co-locate leaders with its workers in the same topology domain, we divide the
problem into subproblems:

- Ensure leader and workers end up on the same flavor
- Find topology domain to co-locate leader and workers

#### Ensure leader and workers end up on the same flavor

> To ensure leader and workers are assigned the same flavor a notion of PodSetGroup is introduced:
> 
> ```golang
> type PodSetTopologyRequest struct {
>   // ...
> 
> 	// PodSetGroup indicates the name of the group of PodSets to which this PodSet belongs to.
> 	// PodSets with the same `PodSetGroup` should be assigned the same ResourceFlavor
> 	//
> 	// +optional
> 	PodSetGroup *string `json:"podSetGroup,omitempty"`
>   
>   // ...
> }
> ```

Reviewer 2: ```suggestion
To ensure leader and workers are assigned the same flavor a notion of PodSetGroup is 
introduced under the PodSetTopologyRequest's PodSetGroupName field (see [Internal APIs](Internal APIs)):
```
To avoid duplication
